docs/rhel4edge_iso.md

```diff
@@ -0,0 +1,212 @@
+```bash
+./scripts/image-builder/build.sh -pull_secret_file ~/pull-secret.txt
+```
+The script performs the following tasks:
+- Check for minimum 10GB of available disk space
+- Set up local MicroShift RPM repository using `packaging/rpm/_rpmbuild` directory contents as a source
+- Set up local OpenShift RPM repository using public `rhocp-${OCP_VERSION}-for-rhel-8-${ARCH}-rpms` repository necessary for CRI-O and OpenShift client package installation
+- Configure Image Builder to use the local MicroShift and OpenShift RPM repositories for image builds
+- Run Image Builder to create an edge container image containing all the MicroShift binaries and dependencies
+- Start a local `ostree` server with the above image
+- Run Image Builder to create an edge installer image comprised of RHEL for Edge OS and MicroShift binaries
+- Rebuild the installer image with the `kickstart.ks` file for performing various OS setup when the host is first booted
+- Perform partial cleanup procedure to reclaim cached disk space
+
+The artifact of the build is the `scripts/image-builder/_builds/microshift-installer.${ARCH}.iso` bootable RHEL for Edge OS image.
+
+### Offline Containers
+The `scripts/image-builder/build.sh` script supports a special mode for including user-specific RPM files into the generated ISO. The remainder of this section demonstrates how to generate container image RPMs for MicroShift and include them in the installation ISO.
+
+When the container images required by MicroShift are preinstalled, `CRI-O` does not attempt to pull them when MicroShift service is first started, saving network bandwidth and avoiding external network connections.
+>If user workloads depend on additional container images, the respective RPMs need to be created by the user separately.
+
+Start by running the `packaging/rpm/make-microshift-images-rpm.sh` script without arguments to see its usage.
+```bash
+$ ./packaging/rpm/make-microshift-images-rpm.sh
+Usage:
+   make-microshift-images-rpm.sh rpm  <pull_secret> <architectures> <rpm_mock_target>
+   make-microshift-images-rpm.sh srpm <pull_secret> <architectures>
+   make-microshift-images-rpm.sh copr <pull_secret> <architectures> <copr_repo>
+
+pull_secret:     Path to a file containing the OpenShift pull secret
+architectures:   One or more RPM architectures
+rpm_mock_target: Target for building RPMs inside a chroot (e.g. 'rhel-8-x86_64')
+copr_repo:       Target Fedora Copr repository name (e.g. '@redhat-et/microshift-containers')
+
+Notes:
+ - The OpenShift pull secret can be downloaded from https://console.redhat.com/openshift/downloads#tool-pull-secret
+ - Use 'x86_64:amd64' or 'aarch64:arm64' as an architecture value
+ - See /etc/mock/*.cfg for possible RPM mock target values
+```
+
+Run the script in the `rpm` mode to pull the images required by MicroShift and generate the RPMs including those image data.
+```bash
+./packaging/rpm/make-microshift-images-rpm.sh rpm ~/pull-secret.txt x86_64:amd64 rhel-8-x86_64
```

Comment:
This script is failing with a permission error. It looks like the pull secret isn't being configured properly because I had to install the pull secret into the root user's account by hand.

```
$ ./packaging/rpm/make-microshift-images-rpm.sh rpm ~/pull-secret.txt x86_64:amd64 rhel-8-x86_64 2>&1 | tee build.log
Trying to pull quay.io/openshift-release-dev/ocp-v4.0-art-dev@sha256:ba44dead03ea74107f90d58525106fb52d27a120b73c6cc8e2be31d37043ca1c...
Error: initializing source docker://quay.io/openshift-release-dev/ocp-v4.0-art-dev@sha256:ba44dead03ea74107f90d58525106fb52d27a120b73c6cc8e2be31d37043ca1c: reading manifest sha256:ba44dead03ea74107f90d58525106fb52d27a120b73c6cc8e2be31d37043ca1c in quay.io/openshift-release-dev/ocp-v4.0-art-dev: unauthorized: access to the requested resource is not authorized
> mkdir -p /home/dhellmann/devel/microshift/microshift/packaging/rpm/_rpmbuild/SPECS
> sudo mkdir -p /tmp/containers/microshift-containers/x86_64
pulling quay.io/openshift-release-dev/ocp-v4.0-art-dev@sha256:ba44dead03ea74107f90d58525106fb52d27a120b73c6cc8e2be31d37043ca1c for arch amd64, to /tmp/containers/microshift-containers/x86_64
> sudo podman pull --arch amd64 --root "/tmp/containers/microshift-containers/x86_64" quay.io/openshift-release-dev/ocp-v4.0-art-dev@sha256:ba44dead03ea74107f90d58525106fb52d27a120b73c6cc8e2be31d37043ca1c
command failed
```

Reply:
It may be a different problem. Are you using up-to-date version of paack.py script? You should see the authfile argument in podman invocations. You should you testing this functionality in the [rpm_images_pull_request](https://github.com/ggiguash/microshift/tree/rpm_images_pull_request) branch.
`
sudo podman pull --authfile /home/microshift/pull-secret.txt --arch amd64 --root "/tmp/containers/microshift-containers/x86_64" quay.io/openshift-release-dev/ocp-v4.0-art-dev@sha256:ba44dead03ea74107f90d58525106fb52d27a120b73c6cc8e2be31d37043ca1c
`